File: easy/ValidAnagram.js
// Given two strings s and t, return true if t is an anagram of s, and false otherwise.

// An Anagram is a word or phrase formed by rearranging the obj[s[i]]s of a different word or phrase, typically using all the original obj[s[i]]s exactly once.

var isAnagram = function(s, t) {
  if (s.length !== t.length) {
    return false
  }
  for (let i = 0; i < s.length; i++) {
    for (let j = 0; j < t.length; j++) {
      if (s[i] === t[j]) {
        t = t.split('');
        t.splice(j, 1);
        t = t.join('');
        break
      }
    }
  }
  return t.length === 0
};

let testStr1 = 'aacc';
let testStr2 = 'ccac';

console.log(isAnagram(testStr1, testStr2))

var isAnagram2 = function(s, t) {
  let obj = {};
  if (s.length !== t.length) {
    return false;
  }
  for (let i = 0; i < s.length; i++) {
    if (obj[s[i]] === undefined) {
      obj[s[i]] = 1;
    } else {
      obj[s[i]]++;
    }
  }
  for (let i = 0; i < t.length; i++) {
    if (obj[t[i]] === undefined) {
      return false;
    } else if (obj[t[i]] === 1) {
      delete obj[t[i]];
    } else {
      obj[t[i]]--;
    }
  }
  return Object.keys(obj).length === 0
}


console.log(isAnagram2(testStr1, testStr2))

File: easy/ValidAnagram.py
# Given two strings s and t, return true if t is an anagram of s, and false otherwise.

# An Anagram is a word or phrase formed by rearranging the obj[s[i]]s of a different word or phrase, typically using all the original obj[s[i]]s exactly once.

def isAnagram(s, t) :
  if len(s) != len(t) :
    return False
  t = list(t)
  for let1 in s :
    i = 0
    for let2 in t :
      if let1 == let2 :
        del t[i]
        break
      i += 1
  return len(t) == 0

print(isAnagram('roxas', 'sorax'))

def isAnagram2(s, t) :
  if len(s) != len(t) :
    return False
  (objS, objT) = ({}, {})
  for letter in s :
    objS[letter] = objS.get(letter, 0) + 1
  for letter in t :
    objT[letter] = objT.get(letter, 0) + 1
  for letter in objS :
    try :
      if (objS[letter] != objT[letter]) :
        return False
    except:
      return False
  return True

print(isAnagram2('roxas', 'soraxn'))


# below is after I figured out that you could directly compare dictionaries in Python
def isAnagram3(s, t) :
  (objS, objT) = ({}, {})
  for letter in s :
    objS[letter] = objS.get(letter, 0) + 1
  for letter in t :
    objT[letter] = objT.get(letter, 0) + 1
  return objS == objT